### ISSUE TITLE
More optimistic results from carettrainmodel$results then from confusionMatrix(model$pred$pred,model$pred$obs)

### ISSUE THREAD

**Issue by t31415** [topepo/caret#897]
I noticed that the results in my models obtained from caret:train , as in

fitCtrl<-trainControl(method='repeatedcv',num=10,repeats=5,savePredictions=TRUE,classProbs=TRUE,summaryFunction=BalAc)
model<-train(Y~.,data=data1,metric='OptBalAcc',method='qda',trControl=fitCtrl)
model$results

are much more optimistic than what I get from

confusionMatrix(model$pred$pred,model$pred$obs)

, where I had expected these to be the same. Or at least for the latter (which I thought were the predictions made by the best model when it is applied again to all of the folds together?) to be at least as good as the former (which I thought is the mean of the holdout performances for each fold/repeat and thus a little less subject to overfitting due to being held out each time when the fit takes place?)

So I'm trying to understand how its possible that Confusionmatrix is always the more pessimistic one, when it is just the training error made when the same best model is used to predict all the data? Or is this not the case? Thanks

**Comment by topepo** [comment]
Providing a reproducible example and the results of `sessionInfo` will help get your question answered.

**Comment by t31415** [comment]
Hi, I've tried to make a simplified example but in the example dataset I found for imbalanced data I'm not able to reprocude the differences of 10% in balanced accuracy I get in my own dataset. The dataset I was working with is rather big and a lot of code was indirectly involved in the custom metrics etc, so I tried to come up with a simple example, but here the difference is a lot smaller. Here is a minimal example:

install.packages('imbalance')
library(imbalance)
#define a custom metric and its traincontrol:
BalAc<-function(data,lev=NULL,model=NULL,b=NULL){
  #just for compatibility with other we keep a useless b=.. argument
  sensvals<-performance(prediction(data$positive,data$obs),'sens')@y.values[[1]]
  specvals<-performance(prediction(data$positive,data$obs),'spec')@y.values[[1]]
  balacvals<-(sensvals+specvals)/2
  optcut<-performance(prediction(data$positive,data$obs),'sens')@x.values[[1]][which.max(balacvals)]
  optBA<-balacvals[which.max(balacvals)]
  out<-c(optBA,optcut)
  names(out) <- c("OptBalAcc", "OptCutoff")
  out
}
trainControlSumfun0<-function(method,number,repeats=NULL,sumfun,...){
  trainControl(method = method,
               number = number,repeats=repeats,savePredictions='final',
               classProbs=TRUE,sampling='smote',
               summaryFunction = function(data,lev=NULL,model=NULL){ sumfun(data,lev,model,...)})
}

fitControl<-trainControlSumfun0(method = 'repeatedcv',number = 10,repeats=5,
                                sumfun = BalAc,b=2) 

#Train a model
trainobj<-train(Class ~., data=newthyroid1,
                method='mda',metric='OptBalAcc',maximize=TRUE,trControl=fitControl,b=2)

#I would think that the following two would be roughly the same, or that at least the former would be less optimistic than the latter, but often it is the other way round:
trainobj$results
confusionMatrix(trainobj$pred$pred,trainobj$pred$obs)

#(Here the difference was only 2%, but on my own dataset I often had much larger differences of up to 10%.) Does that difference mean something is wrong, or is it possible for the latter to be much less good than the former, and why is that?  am I right in assuming that the former is a CV error, ie the average of all hold out errors, while the latter is a pure training error as it doesnt hold out the data it predicts and should therefore yield a higher (apparent) performance due to more overfitting?
Thanks,


PS: the output of 'sessioninfo' is the following:

function (package = NULL) 
{
    z <- list()
    z$R.version <- R.Version()
    z$platform <- z$R.version$platform
    if (nzchar(.Platform$r_arch)) 
        z$platform <- paste(z$platform, .Platform$r_arch, sep = "/")
    z$platform <- paste0(z$platform, " (", 8 * .Machine$sizeof.pointer, 
        "-bit)")
    z$locale <- Sys.getlocale()
    if (.Platform$OS.type == "windows") {
        z$running <- win.version()
    }
    else if (nzchar(Sys.which("uname"))) {
        uname <- system("uname -a", intern = TRUE)
        os <- sub(" .*", "", uname)
        z$running <- switch(os, Linux = if (file.exists("/etc/os-release")) {
            tmp <- readLines("/etc/os-release")
            t2 <- if (any(startsWith(tmp, "PRETTY_NAME="))) sub("^PRETTY_NAME=", 
                "", grep("^PRETTY_NAME=", tmp, value = TRUE)[1L]) else if (any(startsWith(tmp, 
                "NAME"))) sub("^NAME=", "", grep("^NAME=", tmp, 
                value = TRUE)[1L]) else "Linux (unknown distro)"
            sub("\"(.*)\"", "\\1", t2)
        } else if (file.exists("/etc/system-release")) {
            readLines("/etc/system-release")
        }, Darwin = {
            ver <- readLines("/System/Library/CoreServices/SystemVersion.plist")
            ind <- grep("ProductUserVisibleVersion", ver)
            ver <- ver[ind + 1L]
            ver <- sub(".*<string>", "", ver)
            ver <- sub("</string>$", "", ver)
            ver1 <- strsplit(ver, ".", fixed = TRUE)[[1L]][2L]
            sprintf("%s %s %s", ifelse(as.numeric(ver1) < 12, 
                "OS X", "macOS"), switch(ver1, `6` = "Snow Leopard", 
                `7` = "Lion", `8` = "Mountain Lion", `9` = "Mavericks", 
                `10` = "Yosemite", `11` = "El Capitan", `12` = "Sierra", 
                `13` = "High Sierra", ""), ver)
        }, SunOS = {
            ver <- system("uname -r", intern = TRUE)
            paste("Solaris", strsplit(ver, ".", fixed = TRUE)[[1L]][2L])
        }, uname)
    }
    if (is.null(package)) {
        package <- grep("^package:", search(), value = TRUE)
        keep <- vapply(package, function(x) x == "package:base" || 
            !is.null(attr(as.environment(x), "path")), NA)
        package <- .rmpkg(package[keep])
    }
    pkgDesc <- lapply(package, packageDescription, encoding = NA)
    if (length(package) == 0) 
        stop("no valid packages were specified")
    basePkgs <- sapply(pkgDesc, function(x) !is.null(x$Priority) && 
        x$Priority == "base")
    z$basePkgs <- package[basePkgs]
    if (any(!basePkgs)) {
        z$otherPkgs <- pkgDesc[!basePkgs]
        names(z$otherPkgs) <- package[!basePkgs]
    }
    loadedOnly <- loadedNamespaces()
    loadedOnly <- loadedOnly[!(loadedOnly %in% package)]
    if (length(loadedOnly)) {
        names(loadedOnly) <- loadedOnly
        pkgDesc <- c(pkgDesc, lapply(loadedOnly, packageDescription))
        z$loadedOnly <- pkgDesc[loadedOnly]
    }
    z$matprod <- as.character(options("matprod"))
    es <- extSoftVersion()
    z$BLAS <- as.character(es["BLAS"])
    z$LAPACK <- La_library()
    class(z) <- "sessionInfo"
    z
}
<bytecode: 0x1177066e0>
<environment: namespace:utils>

**Comment by t31415** [comment]
Oh I think I realised the stupid error I made, it will of course run through sens and spec in opposite order, as it runs to the cutoffs in a same order and 1-spec is plotted..  So I should have maximized sensvals+rev(specvals) in my custom metric of course..

**Comment by t31415** [comment]
The cutoffs used are in the same order apparently after all, so then I think the above would have been the correct balanced accuracy after all, so I really dont get why it would be so much higher in CV results than the one in ConfusionMatrix..

**Comment by topepo** [comment]
`trainControl` will not pass arguments into the summary function (and `b` isn't used inside of the function). To be honest, you've made this really more complex than it probably needs to be. 

> Does that difference mean something is wrong, or is it possible for the latter to be much less good than the former, and why is that? 

So, to try to answer your question, the difference could really be driven by the dataset size. Since there is an imbalance, the number of events is  probably pretty low so that the variability in the sensitivity is probably pretty high (so then the _J_ statistic that you're calculating will have high variance too). 

> am I right in assuming that the former is a CV error, ie the average of all hold out errors, while the latter is a pure training error as it doesn't hold out the data it predicts and should therefore yield a higher (apparent) performance due to more overfitting?

`train` reports the average of the statistics computed inside the iterations of resampling and your estimate using `confusionMatrix` uses the same predictions but pools the data. I would normally say that the latter approximates the former (but they would not be the same). I've done the same thing to show the out-of-sample data to show an ROC curve to try to "approximate" the curve being estimated by resampling. 

> I'm trying to understand how its possible that `confusionMatrix` is always the more pessimistic one,

I would have thought that it would be more optimistic since it is pooling a lot of data together to make the estimate. 

Try mocking up an example for testing. `twoClassSim()` has an option for the intercept and that can be used to modulate the event rate. Using a simulated data set would help because you could also simulate a very large test set to approximate the main component of error of the model.

**Comment by t31415** [comment]
Hi Max, thanks for your reply. I think there was actually also another problem with my code, I had really neglected something much more basic I think, but I think now I understand what error (or at least one of the errors) I made in trying to use custom metrics (I should have mentioned it here afterwards but I still wasnt 100% sure and I was also in the middle of an exam session so i didnt come to doing that yet). Maybe you can confirm if it makes sense that the following could be the reason for the discrepancy in the opposite sense (ie more pessimistic instead of optimistic), or whether its still not correct:

You see, all of the custom metrics where I had noticed this problem, they were all cases where I wanted to use an optimal cutoff value. So for example not just the F-b measure or Balanced Accuracy at cutoff 0.5, but at that cutoff where it becomes maximal.  But I think now I overlooked the fact that caret will (probably?) be optimizing this cutoff in every repeat and for every holdout fold separately, causing it to use different cutoffs for different holdout folds and for different repeats, which is of course not going to mimic realistic performance on the test set because you can only use one single predetermined cutoff there.. 

And then when I use the pooled predictions, I could optimize only once a single cutoff that is optimal for the whole pooled group, so it yields of course a much lower (but much more realistic) performance. 
Do you think this explanation of the difference makes sense? 

So what I should have done probably was to use a fixed cutoff in that summary function for the custom metric, and then afterwards run an external loop over a few of these cutoffs to select the the value where the performance is best, no? I thought caret would do it like this but in retrospect it was of course doing this within every other holdout fold and every other repeat no, instead of once for the whole bunch?

So I wonder a bit now if there would have been an easy way to include that cutoff as an extra tuning parameter in caret? 

(So for the poor question from originally by the way, it was the first time i'm using the package (and its really incredibly usefull btw, nice work!) but I still didnt understand well enough how it worked apparently). Best regards and thanks again for your comments, Tim

**Comment by topepo** [comment]
About

> So what I should have done probably was to use a fixed cutoff in that summary function for the custom metric, and then afterwards run an external loop over a few of these cutoffs to select the the value where the performance is best, no?

and

> So I wonder a bit now if there would have been an easy way to include that cutoff as an extra tuning parameter in caret?

I wrote an [example of some custom code](https://topepo.github.io/caret/using-your-own-model-in-train.html#illustrative-example-5-optimizing-probability-thresholds-for-class-imbalances) to optimize the threshold based on metrics during resampling. That method is a little complex though and gets more complex when the model uses the "sub-model trick".  

Instead, I wrote a function that can do the cutoff analysis after the model has been fit called [`thresholder`](https://www.rdocumentation.org/packages/caret/versions/6.0-80/topics/thresholder). That would probably work well for what you are trying to do (or, if not, you could adapt it). 

> I thought caret would do it like this but in retrospect it was of course doing this within every other holdout fold and every other repeat no, instead of once for the whole bunch?

Yes, but you could pick the best threshold using resampling (and understand the variation around that value) then set it for the final model as we would with other tuning parameters.

**Comment by t31415** [comment]
Hi Max, that example and the thresholder function indeed seem to be doing exactly what I was referring to here, thanks a lot for your suggestions, and best regards!
